tests/test_projectbuilder.py

```diff
@@ -570,4 +576,54 @@ def test_log(mocker, caplog, test_flit_path):
         ('INFO', 'something'),
     ]
     if sys.version_info >= (3, 8):  # stacklevel
-        assert [(record.lineno) for record in caplog.records] == [305, 305, 338, 368, 368, 562]
+        assert caplog.records[-1].lineno == 562
+
+
+@pytest.mark.parametrize(
+    ['pyproject_toml', 'raises'],
+    [
+        (
+            {'build-system': {}},
+            pytest.raises(build.BuildSystemTableValidationError, match='`requires` is a required property'),
+        ),
+        (
+            {'build-system': {'requires': 'not an array'}},
+            pytest.raises(build.BuildSystemTableValidationError, match='`requires` must be an array of strings'),
+        ),
+        (
+            {'build-system': {'requires': [1]}},
+            pytest.raises(build.BuildSystemTableValidationError, match='`requires` must be an array of strings'),
+        ),
+        (
+            {'build-system': {'requires': ['foo']}},
+            nullcontext(),
+        ),
+        (
+            {'build-system': {'requires': ['foo'], 'build-backend': ['not a string']}},
+            pytest.raises(build.BuildSystemTableValidationError, match='`build-backend` must be a string'),
+        ),
+        (
+            {'build-system': {'requires': ['foo'], 'build-backend': 'bar'}},
+            nullcontext(),
+        ),
+        (
+            {'build-system': {'requires': ['foo'], 'backend-path': 'not an array'}},
+            pytest.raises(build.BuildSystemTableValidationError, match='`backend-path` must be an array of strings'),
+        ),
+        (
+            {'build-system': {'requires': ['foo'], 'backend-path': [1]}},
+            pytest.raises(build.BuildSystemTableValidationError, match='`backend-path` must be an array of strings'),
+        ),
+        (
+            {'build-system': {'requires': ['foo'], 'build-backend': 'bar', 'backend-path': ['baz']}},
+            nullcontext(),
+        ),
+        (
+            {'build-system': {'requires': ['foo'], 'unknown-prop': False}},
+            pytest.raises(build.BuildSystemTableValidationError, match='Unknown properties: unknown-prop'),
+        ),
+    ],
+)
+def test_parse_build_system_table_type_validation(pyproject_toml, raises):
+    with raises:
+        build._parse_build_system_table(pyproject_toml)
```

Comment:
I would rather instantiate a `ProjectBuilder` here.

Reply:
I'd have to serialise them all as TOML, create a tmp dir for each and write them on disk - I don't think I can disentangle the disk I/O with the mocker.

Reply:
I think you can, as you should be able to patch `builtins.open`, but I'll leave that up to you. This is just a nitpick.

Reply:
I think I might leave that for another day then 😛